rocAL/source/loaders/audio/audio_loader.cpp

```diff
@@ -0,0 +1,282 @@
+    _circ_buff.unblock_writer();
+    _circ_buff.reset();
+    if (_load_thread.joinable())
+        _load_thread.join();
+}
+
+void AudioLoader::initialize(ReaderConfig reader_cfg, DecoderConfig decoder_cfg, RocalMemType mem_type, unsigned batch_size, bool decoder_keep_original) {
+    if (_is_initialized)
+        WRN("initialize() function is already called and loader module is initialized")
+    if (_output_mem_size == 0)
+        THROW("output audio size is 0, set_output() should be called before initialize for loader modules")
+    _mem_type = mem_type;
+    _batch_size = batch_size;
+    _loop = reader_cfg.loop();
+    _audio_loader = std::make_shared<AudioReadAndDecode>();
+    size_t shard_count = reader_cfg.get_shard_count();
+    int device_id = reader_cfg.get_shard_id();
+    try {
+        // set the device_id for decoder same as shard_id for number of shards > 1
+        if (shard_count > 1) {
+            _audio_loader->Create(reader_cfg, decoder_cfg, _batch_size, device_id);
+        } else {
+            _audio_loader->Create(reader_cfg, decoder_cfg, _batch_size);
+        }
+    } catch (const std::exception& e) {
+        de_init();
+        throw;
+    }
+    _max_decoded_samples = _output_tensor->info().max_shape().at(0);
+    _max_decoded_channels = _output_tensor->info().max_shape().at(1);
+    _decoded_audio_info._data_names.resize(_batch_size);
+    _decoded_audio_info._original_audio_samples.resize(_batch_size);
+    _decoded_audio_info._original_audio_channels.resize(_batch_size);
+    _decoded_audio_info._original_audio_sample_rates.resize(_batch_size);
+    _circ_buff.init(_mem_type, _output_mem_size, _prefetch_queue_depth);
+    _is_initialized = true;
+    LOG("Loader module initialized");
+}
+
+void AudioLoader::start_loading() {
+    if (!_is_initialized)
+        THROW("start_loading() should be called after initialize() function is called")
+
+    _remaining_audio_count = _audio_loader->Count();
+    _internal_thread_running = true;
+    _load_thread = std::thread(&AudioLoader::load_routine, this);
+}
+
+LoaderModuleStatus
+AudioLoader::load_routine() {
+    LOG("Started the internal loader thread");
+    LoaderModuleStatus last_load_status = LoaderModuleStatus::OK;
+    // Initially record number of all the audios that are going to be loaded, this is used to know how many still there
+
+    while (_internal_thread_running) {
+        auto data = reinterpret_cast<float*>(_circ_buff.get_write_buffer());
+        if (!_internal_thread_running)
+            break;
+
+        auto load_status = LoaderModuleStatus::NO_MORE_DATA_TO_READ;
+        {
+            load_status = _audio_loader->Load(data,
+                                              _decoded_audio_info._data_names,
+                                              _max_decoded_samples,
+                                              _max_decoded_channels,
+                                              _decoded_audio_info._original_audio_samples,
```

Comment:
load routine can just take a pointer to decoded_audio_info and access all other members from that